rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,337 @@
+persistent state. A lot of things wouldn't change, e.g., starting a node:
+
+```bash
+vast -e 1.2.3.4 start
+```
+
+There are of course convenience defaults we can use, like assuming `from -` and
+`to -` being the default. This would make the above look like this:
+
+```bash
+vast exec 'where EXPRESSION |
+           write json
+```
+
+Side effect: we would get our envisioned convert operator almost for free:
+
+```bash
+vast exec 'read zeek | print csv'
+# Obviates the need for:
+vast exec 'convert zeek csv'
+```
+
+### Node Interaction
+
+When considering pipelines in conjunction with VAST nodes, we may want to
+improve the UX of working with pipelines. In the new model, VAST would be
+an `Arrow` source or sink, making the primary interaction on the data plane
+through `push` and `pull` operators.
+
+Making `push` and `pull` top-level commands would streamline UX. For example,
+consider this `vast export` pipeline:
+
+```bash
+# exec
+vast exec 'pull vast://1.2.3.4 |
+           where EXPRESSION |
+           write json'
+# pull
+vast pull 'where EXPRESSION |
+           write json
+```
+
+This assumes that `pull` tries the default VAST endpoint of 1.2.3.4 to connect.
+
+An `import` degenerates to:
+
+```bash
+# exec
+vast exec 'read zeek | where EXPRESSION | push vast:///'
+# push
+vast push 'read zeek | where EXPRESSION'
+```
+
+Note that this hoisting into commands only works because `pull` and `push` are
+head/tail of the pipeline.
+
+### Mutating existing data
+
+Security telemetry data is often messy. After the data collection aspects have
+been sorted out and analysts interact with the data, it becomes clear that some
+part of the data needs to be cleaned. This matches ELT mindset of dumping
+everything first so that the data is there, and then tweak as you go.
+
+Example operations include, migrating data to a different schema, enriching
+telemetry (e.g., with GeoIP data), and fixing invalid entries.
+
+Mutation of data is non-trivial in that it requires exclusive access to the
+data. Otherwise data races may occur. Consequently, we need transactional
+interface (ACID) to support this operation.
+
+As long as VAST owns the underlying data (i.e., only VAST is allowed to make
+changes), VAST can already mutate of data at rest. [Spatial
+compaction][mutate-at-rest] uses a pre-defined disk quota as trigger to apply
+pipelines to a subset of to-be-transformed data. After VAST applies the
+pipeline, VAST optionally removes the original data in an atomic fashion.
+
+[mutate-at-rest]: https://vast.io/docs/use-vast/transform#modify-data-at-rest
+
+Modeled after the `compaction` plugin, we may consider exposing a mutable
+pipeline interface through a dedicate command. This command interacts with a
+node, and could be seen as a combination of `pull` and `push` that executes
+remotely. For example:
+
+```bash
+vast mutate --types=A,B,C 'where 10.0.0.0/8 | put orig.h resp.h'
```

Comment:
I don't understand the need for the `--types` argument here. We already can just support arbitrary expressions for the lookup, so why not just make it `vast mutate '<expr> | <pipeline>'`?

Reply:
I removed `--types`. Closing this, as I see it resolved.